presto-main/src/main/java/io/prestosql/sql/planner/DomainCoercer.java

```diff
@@ -0,0 +1,244 @@
+import io.prestosql.metadata.ResolvedFunction;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.function.InvocationConvention;
+import io.prestosql.spi.predicate.AllOrNoneValueSet;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.Marker;
+import io.prestosql.spi.predicate.Range;
+import io.prestosql.spi.predicate.Ranges;
+import io.prestosql.spi.predicate.SortedRangeSet;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeOperators;
+import io.prestosql.sql.InterpretedFunctionInvoker;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.prestosql.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.prestosql.spi.function.OperatorType.SATURATED_FLOOR_CAST;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Apply saturated floor casts for implicit coercions on TupleDomain.
+ * This class does not handle Float.NaN and Double.NaN because
+ * currently it is used only in dynamic filtering where NaNs are not part of TupleDomain.
+ */
+public final class DomainCoercer
+{
+    private DomainCoercer() {}
+
+    public static Domain applySaturatedCasts(
+            Metadata metadata,
+            TypeOperators typeOperators,
+            Session session,
+            Domain domain,
+            Type coercedValueType)
+    {
+        return new ImplicitCoercer(metadata, typeOperators, session, domain, coercedValueType).applySaturatedCasts();
+    }
+
+    private static class ImplicitCoercer
+    {
+        private final ConnectorSession connectorSession;
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final ResolvedFunction saturatedFloorCastOperator;
+        private final ResolvedFunction castToOriginalTypeOperator;
+        private final MethodHandle comparisonOperator;
+        private final Domain domain;
+        private final Type coercedValueType;
+
+        private ImplicitCoercer(Metadata metadata, TypeOperators typeOperators, Session session, Domain domain, Type coercedValueType)
+        {
+            this.connectorSession = requireNonNull(session, "session is null").toConnectorSession();
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata);
+            this.domain = requireNonNull(domain, "domain is null");
+            this.coercedValueType = requireNonNull(coercedValueType, "coercedValueType is null");
+            Type originalValueType = domain.getType();
+            try {
+                this.saturatedFloorCastOperator = metadata.getCoercion(SATURATED_FLOOR_CAST, originalValueType, coercedValueType);
```

Comment:
There are two possible approaches: SATURATED_FLOOR_CAST and coerce-coerce (as in `UnwrapCastInComparison`).
What was the reason to choose the former over the latter?
Can we capture the reasoning in writing somewhere?

cc @sopel39 @martint

Reply:
ping?